### PR TITLE
Ml29 adv validation

### LINKED ISSUE #29: Add Adversarial validation check on the stage of *scale_data*
Implement an adversarial validation checks to avoid a common problem when test/train data are not coming from the same population.

**See more info here:**
http://fastml.com/adversarial-validation-part-two/

**Possible implementation:**
https://www.kaggle.com/konradb/adversarial-validation-and-other-scary-terms

### PR DESCRIPTION
**Pull request by atselikov**
Added adversarial validation checking on the stage of data loading, so it will inform user in case of train and test is completely different.

resolve #29

### CODE REVIEW

File: preprocessing/data_science_help_functions.py [hunk 1]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+
+    :param DataFrame df_joined: Feature dataframe
+    :param Series y: Target series
+
+    :return:
+            | clf: Trained model
+            | mean of KFold validation results (ROC-AUC scores)
+    """
+
+    skf = StratifiedKFold(n_splits=3, shuffle=True, random_state=44)
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: Here is a very rigid parameter definition. I don't know is this work for all cases. I would suggest printing out the xgb_params to the user or to inform him/her about the configuration and tell how to change this. I would suggest extracting this to an external json file and pointing to it and loading using json.load(filename)

File: preprocessing/data_science_help_functions.py [hunk 2]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    xgb_params = {
+        'learning_rate': 0.1, 'max_depth': 4, 'subsample': 0.9,
+        'colsample_bytree': 0.9, 'objective': 'binary:logistic',
+        'silent': 1, 'n_estimators': 50, 'gamma': 1,
+        'min_child_weight': 4
+    }
+    clf = xgb.XGBClassifier(**xgb_params, seed=10)
+    results = []
+    logger.info('Adversarial validation checking:')
+    for fold, (train_index, test_index) in enumerate(skf.split(df_joined, y)):
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: I would suggest extracting this in a new function if it is possible

File: preprocessing/data_science_help_functions.py [hunk 3]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+        fold_score = roc_auc_score(fold_yval, fold_ypred)
+        results.append(fold_score)
+        logger.info(f"Fold: {fold + 1} shape: {fold_xtrain.shape} score: {fold_score}")
+
+    return round(np.mean(results), 2), clf
+
+
+def xgb_important_features(xgb: xgb.sklearn.XGBClassifier,
+                           top_features: int = 5) -> str:
+    """ Get top of the most important features from a trained model
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: I would suggest using a short title: important parameters extractor

File: preprocessing/data_science_help_functions.py [hunk 4]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+
+    :param XGBClassifier xgb: A trained model
+    :param int top_features: Max length of features to send back
+
+    :return:
+            | A string with a list of the most important features plus their importance
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: I would suggest using -<space> .. instead |

File: preprocessing/data_science_help_functions.py [hunk 5]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    """
+
+    # get features
+    feat_imp = xgb.get_booster().get_score(importance_type='gain')
+
+    # round importances
+    for dict_key in feat_imp:
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: I would suggest extracting it in a new small function

File: preprocessing/data_transformer.py [hunk 6]
```diff
@@ -266,7 +266,7 @@ def clean_categorical_features(dataframe_dict: object,
     # Checking cycle
     if print_results:
         print('*' * 10)
-        print(f"Checking the difference in categorical columns in {dataframe_dict.keys()[0]} and {dataframe_dict.keys()[1]} datasets: ")
+        print(f"Checking the difference in categorical columns in {list(dataframe_dict.keys())[0]} and {list(dataframe_dict.keys())[1]} datasets: ")
```
Comment (mahedeeb) [inline preprocessing/data_transformer.py]: maybe for two keys, the list could not be a problem, but I would suggest following the best practice as Ahmed suggested to me before:
next(iter(dataframes_dict.keys()))

Comment (atselikov) [reply preprocessing/data_transformer.py]: For some reason the previous variant (without a list) doesn't for me at all any more.
just checked `next(iter(dataframe_dict.keys())` - it works - will replace to it

File: preprocessing/data_science_help_functions.py [hunk 7]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    This function tries to check whether test and train data coming from the same data distribution.
+
+    :param dict dataframes_dict: A dictionary that contains pandas dataframes e.g. dataframes_dictionary ={
+                'train': train_dataframe, 'test': test_dataframe}
+    :param int max_dataframe_length: Max length of dataframe to be considered - make adversarial validation faster
+    :param list ignore_columns: List of column to ignore (ID, target, etc...)
+    :param float threshold: A value larger than 0 and less than 1. If the result of calculation is greater than threshold - there is sugnificant difference between train and test data
+
+    :return:
+            | adversarial_validation_result: Adversarial validation score.
```
[the remark this answers is not in the document]
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: fixed

File: flows/flows.py [hunk 8]
```diff
@@ -185,10 +186,14 @@ def load_data(self, path: str, files_list: list, rows_amount: int = 0) -> Tuple[
         print(term.green_on_black("average_score, is_outlier = flow.outliers_extractor"
                                   "(dataframe_dict: dict, key_i: str)"))
 
-        _, _, possible_problems = detect_id_target_problem(dataframes_dict)
+        possible_ids, possible_target, possible_problems = detect_id_target_problem(dataframes_dict)
 
         _ = check_if_target_columns_are_imbalanced(dataframes_dict, possible_problems, self.kl_div_threshold)
 
+        # Check adversarial validation
+        ignore_columns = list(set(possible_ids).union(set(possible_target)))
```
[the remark this answers is not in the document]
Comment (atselikov) [reply flows/flows.py]: ok, created issue 

> adversarial_validation(dataframes_dict, ignore_columns)
> I would suggest adding a new list of dataframes names. Maybe the user has three 3 dataframes(datasets)[trian, validate, test] and the user interested in applying the test between validate and test, the user should be able to do this by applying the following:
> adversarial_validation(dataframes_dict, ignore_columns, datasets=["validate", "test"]) or
> adversarial_validation(dataframes_dict, ignore_columns, datasets=["train", "test"]) or
> adversarial_validation(dataframes_dict, ignore_columns, datasets=["train", "validate"])

created #43 for it

[the remark this answers is not in the document]
Comment (atselikov) [reply flows/flows.py]: > I would suggest keeping this (ignore_columns) a set. I discussed with Ahmad before sometimes ago and we decided to refactor all ignore_columns from lists to sets and I opened an issue for it #26. This is because it is more efficient.

ok, let's keep it in the #26 scope - no immediate solution here. For now, the only list is accessible...

File: preprocessing/data_science_help_functions.py [hunk 9]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
     print(f"The possible possible_target are:\n {possible_target}")
     print(f"The type of the problem that should be solved:\n {possible_problems}")
     return possible_ids, possible_target, possible_problems
+
+
+def adversarial_validation(dataframe_dict: dict,
+                           ignore_columns: list = [],
+                           max_dataframe_length: int = 100000,
+                           threshold: float = 0.7) -> float:
+    """ Training a probabilistic classifier to distinguish train/test examples.
```
[the remark this answers is not in the document]
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: done

File: preprocessing/data_science_help_functions.py [hunk 10]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    See more info here: http://fastml.com/adversarial-validation-part-one/
+
+    This function tries to check whether test and train data coming from the same data distribution.
```
[the remark this answers is not in the document]
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: ok ,check is better, agree

File: preprocessing/data_science_help_functions.py [hunk 11]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    :param float threshold: A value larger than 0 and less than 1. If the result of calculation is greater than threshold - there is sugnificant difference between train and test data
+
+    :return:
+            | adversarial_validation_result: Adversarial validation score.
+    """
+
+    # Check if it only one dataframe provided
+    if len(dataframe_dict) != 2:
+        # do nothing and return the original data
+        logger.info("Can't apply adversarial_validation because count of dataframes is not equal to 2")
```
[the remark this answers is not in the document]
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: ok

File: preprocessing/data_science_help_functions.py [hunk 12]
```diff
@@ -189,3 +196,140 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    for df in [train, test]:
+        if len(df) > max_dataframe_length:
+            df = df.head(max_dataframe_length)
+
+    # add identifier and combine
+    train['istrain'] = 1
+    test['istrain'] = 0
+    df_joined = pd.concat([train, test], axis=0)
+
+    # convert non-numerical columns to integers
```
[the remark this answers is not in the document]
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: done

File: preprocessing/data_science_help_functions.py (hunk 1 above)
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: Simplified it a bit

File: preprocessing/data_science_help_functions.py (hunk 5 above)
Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: done

File: preprocessing/data_transformer.py (hunk 6 above)
Comment (atselikov) [reply preprocessing/data_transformer.py]: replaced it with labels and iterator

File: preprocessing/data_science_help_functions.py [hunk 13]
```diff
@@ -189,3 +197,178 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+
+    :return:
+            - adversarial_validation_result: Adversarial validation score.
+    """
+
+    # Check if it only one dataframe provided
+    if len(dataframe_dict) != 2:
+        # do nothing and return the original data
+        print("Can't apply adversarial_validation because count of dataframes is not equal to 2")
+        return
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: Thanks a lot for your great refactoring. You have a "return" statement here. I think you forgot removing it. I tried to test your code in one of the flows, but it did not work. I would suggest testing your code in all flows and check if it works without any problem. Even better, I would recommend strongly to write unit tests for the functions that you created.

File: preprocessing/data_science_help_functions.py [hunk 14]
```diff
@@ -189,3 +197,178 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+
+    # TODO: support > 2 dataframes
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: I would suggest removing the "TODO" functions and open issues for those tasks

File: preprocessing/data_science_help_functions.py [hunk 15]
```diff
@@ -189,3 +197,178 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+    :return:
+            - df_joined: Joined dataframe.
+    """
+
+    if len(ignore_columns) > 0:
+        columns_to_use = [x for x in list(test.columns) if x not in ignore_columns]
+        train = train[columns_to_use]
+        test = test[columns_to_use]
+    # add identifier and combine
+    train['istrain'] = 1
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: Could you please check this again. do you need to repeat the code here. Let us discuss this if the answer is yes.

File: preprocessing/data_science_help_functions.py (hunk 1 above)
Comment (mahedeeb) [reply preprocessing/data_science_help_functions.py]: Let us discuss this. I would still write it to an external file

File: preprocessing/data_science_help_functions.py [hunk 16]
```diff
@@ -189,3 +197,178 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
...
+
+    # get features
+    feat_imp = xgb.get_booster().get_score(importance_type='gain')
+
+    sorted_x = round_and_sort_dict(feat_imp)
+
+    return str(list(sorted_x[:top_features]))
+
+
+def round_and_sort_dict(feat_imp: dict) -> dict:
```
Comment (mahedeeb) [inline preprocessing/data_science_help_functions.py]: Pycharm says that this function should return a dict but it got a list. Is the output a list?

Comment (atselikov) [reply preprocessing/data_science_help_functions.py]: string - the purpose to print the first top_features
